elasticsearch/src/main/java/com/tcdi/zombodb/postgres/ZombodbBulkAction.java

```diff
@@ -172,199 +178,134 @@ private RestResponse buildResponse(BulkResponse response, XContentBuilder builde
-                // so lets do that and try one more time
-                client.admin().indices().refresh(Requests.refreshRequest(defaultIndex)).actionGet();
-                retries++;
-                continue;
-            }
-
-            break;
-        }
 
-        if (response.getHits().getHits().length != lookup.size())
-            throw new RuntimeException("Did not find all previous ctids an UPDATE");
-
-        for (SearchHit hit : response.getHits()) {
-            String prevCtid = hit.field("_prev_ctid").getValue();
-            IndexRequest doc = lookup.get(hit.id());
-
-            if (doc == null)
-                continue;
-
-            Map<String, Object> data = doc.sourceAsMap();
-            Number xid = (Number) data.get("_xid");
-
-            data.put("_prev_ctid", prevCtid);
-            doc.source(data);
-            doc.routing(prevCtid);
-            doc.opType(IndexRequest.OpType.CREATE);
-            doc.versionType(VersionType.FORCE);
-            doc.version(xid.longValue());
-
-            trackingRequests.add(
-                    new IndexRequestBuilder(client)
-                            .setId(hit.id())
+            xmaxRequests.add(
+                    new DeleteRequestBuilder(client)
                             .setIndex(defaultIndex)
-                            .setType("state")
-                            .setRouting(prevCtid)
-                            .setOpType(IndexRequest.OpType.INDEX)
-                            .setVersionType(VersionType.FORCE)
-                            .setVersion(xid.longValue())
-                            .setSource("_ctid", prevCtid)
+                            .setType("xmax")
+                            .setRouting(doc.id())
+                            .setId(doc.id())
                             .request()
             );
 
+            doc.routing(doc.id());
         }
-
-        return trackingRequests;
     }
 
-    private List<ActionRequest> handleIndexRequestsUsingPkey(Client client, List<ActionRequest> requests, String defaultIndex, String pkeyFieldname) {
-        List<ActionRequest> trackingRequests = new ArrayList<>();
+    private void handleIndexRequests(Client client, List<ActionRequest> requests, String defaultIndex, int requestNumber, List<ActionRequest> xmaxRequests, List<ActionRequest> abortedRequests) {
+
+        int cnt = 0;
         for (ActionRequest ar : requests) {
             IndexRequest doc = (IndexRequest) ar;
             Map<String, Object> data = doc.sourceAsMap();
-            Object pkey = data.get(pkeyFieldname);
-            Object prevCtid = data.get("_prev_ctid");
-            Number xid = (Number) data.get("_xid");
+            String prev_ctid = (String) data.get("_prev_ctid");
+            Number xmin = (Number) data.get("_xmin");
+            Number cmin = (Number) data.get("_cmin");
+            Number sequence = (Number) data.get("_zdb_seq");    // -1 means an index build (CREATE INDEX)
+
+            {
```

Comment:
TODO:  In a future release, move this tuple encoding to the Postgres side of things so as to avoid manipulating and copying large byte arrays for every row we're indexing.